### ISSUE TITLE
Cannot set log level of `ignite.distributed` because of calls to setup_logger inside auto functions

### ISSUE THREAD

**Issue by schuhschuh** [pytorch/ignite#1763]
## 🐛 Bug description

The `ignite.distributed` auto functions such as `ignite.distributed.auto_dataloader` use `setup_logger` to get a `logging.Logger` instance. However, this function always sets the logging level to `logging.INFO` and furthermore removes ands sets up the logging handlers. Functions should not be setting up a logger object if it is already initialized such that user code can specify a different log level, handlers, or log message format string in their code (e.g., main function).

See for example:
- https://github.com/pytorch/ignite/blob/6d490c8f3083f21ba8f82eb646d9052a06fae3e2/ignite/distributed/auto.py#L67
- https://github.com/pytorch/ignite/blob/6d490c8f3083f21ba8f82eb646d9052a06fae3e2/ignite/distributed/launcher.py#L201

Note that in case of `ignite.distributed.launcher.Parallel` one can modify the `self.logger` instance, but this logger already prints messages right after `setup_logger` in its `__init__` function, hence, making it impossible to adjust the log level, format, or handlers for these first messages also.

## Environment

- PyTorch Version (e.g., 1.4):
- Ignite Version (e.g., 0.3.0):
- OS (e.g., Linux):
- How you installed Ignite (`conda`, `pip`, source):
- Python version:
- Any other relevant information:

**Comment by vfdev-5** [comment]
Thanks for the report @schuhschuh ! 
Yes, this part of code should be definitetly improved and I think @sdesrozis also proposed to do that.

@schuhschuh if you would like help us with improving the code, please feel free to discuss it here and then send a draft PR. Thanks

**Comment by sdesrozis** [comment]
We could add logger argument to functions. Or define a logger for the module `distributed`.

@vfdev-5 other ideas ?

**Comment by vfdev-5** [comment]
Defining a logger for the module could be nice but can be tricky as it should be adapted and re-adapted to the dist config...

**Comment by schuhschuh** [comment]
I would say you already have global logging object for these functions anyway by the use of `logging.getLogger()` in `setup_logger()`. This is what makes it possible to adjust the logger without accessing a module level logger instance reference. Multiple calls to `auto_*()` functions will give you the same logger instance. The issue is just that this instance will also always reset the level, format, and handlers.

Would it work to just use a `setup_logger` variant (maybe a flag for this function) which instead of removing any previous handlers and setting up the logger retrieved by `logging.getLogger(name)` would simply check `logger.hasHandlers()` and just return the already set up logger object if `True` at [this line](https://github.com/pytorch/ignite/blob/6d490c8f3083f21ba8f82eb646d9052a06fae3e2/ignite/utils.py#L142)? One could still check and remove all handlers for `distributed_rank > 0`. But given that these loggers have a `NullHandler` set by any previous call to the `auto_*` functions, I don't think this would be necessary. Also, by only checking `hasHandlers()` and making any changes to the logger object only if `False`, you leave the user the option to setup any desired handlers even for loggers where the `distributed_rank` of the process is greater than zero.

For example:
```
def setup_logger(
    name: Optional[str] = None,
    level: int = logging.INFO,
    stream: Optional[TextIO] = None,
    format: str = "%(asctime)s %(name)s %(levelname)s: %(message)s",
    filepath: Optional[str] = None,
    distributed_rank: Optional[int] = None,
    reset: bool = True,
) -> logging.Logger:
    logger = logging.getLogger(name)
    if logger.hasHandlers() and not reset:
        return logger
    # [...]
```

**Comment by vfdev-5** [comment]
@schuhschuh yes, I agree that we could use `setup_logger` for setting up and retrieving existing logger. 

@sdesrozis what do you think ?

**Comment by sdesrozis** [comment]
Sorry for the lag, I missed this message !!

Just a remark. If user first calls `logging.basicConfig()` (or any configuration), the `root` logger is defined and embeds handlers. So after that, calling `setup_logger()` considering the suggestion of @schuhschuh should return the `root` logger initialized by the user... and it does not take into account the specific ignite handlers pushed in the method.

I'll look into it on my side.

**Comment by sdesrozis** [comment]
Another remark. 

We use native `logging` module in `Engine` and `TerminateOnNan` and `LRFinder` (grep told me).

```python
logging.basicConfig(level=logging.INFO)

engine = Engine(lambda e, b: None)
engine.run([0, 1, 2], max_epochs=2)
```
prints in console
```
INFO:ignite.engine.engine.Engine:Engine run starting with max_epochs=2.
INFO:ignite.engine.engine.Engine:Epoch[1] Complete. Time taken: 00:00:00
INFO:ignite.engine.engine.Engine:Epoch[2] Complete. Time taken: 00:00:00
INFO:ignite.engine.engine.Engine:Engine run complete. Time taken: 00:00:00
```

Do we want that behavior ? IMO it would be better to use `setup_logger` in our modules rather than native `logging` as we have specific handlers for `distributed` mode. However, default level of `setup_logger` is `logging.INFO`. We would be better to have `logging.DEBUG` for internal.

What do you think ?

**Comment by schuhschuh** [comment]
@sdesrozis Actually, setup_logger is using the native logging module. You are not actually retrieving the root logger, because getLogger is called with the module name as argument.The question should only be whether you are able to tell if the specific logger that you obtain, which is not a reference to the root logger, has had handlers set? I would think/hope so.

**Comment by sdesrozis** [comment]
I agree that we always use native logging. I was talking about configuration, sorry it was not understandable.

If you call `setup_logger()` without argument, you get the root logger. I think that is a point to have in mind. `logging.basicConfig(...)` should not conflict with our way to setup loggers. If a logger is setup  with a specific name, I assume that it won't conflict. Finally I suppose it could help if `setup_logger()` returns a logger that differs from root.

What I said may be beside the point of the issue. I understand that multiple calls reset handlers and level. The suggested solution aims to configure once each logger. (Loggers must not have handlers attached, take care of root).

I will have a look on my own.

**Comment by vfdev-5** [comment]
@schuhschuh please take a look at this PR : #1945 with a suggested improvement to fix this problem. Thanks!

**Comment by schuhschuh** [comment]
Hey, this is great! Thanks for working on this and keeping me in the loop. Sorry I didn't provide feedback as I was distracted.

The general recommendation for any Python lib should be to never use the root logger internally. Root is for applications. So I think this works well with how you've been adapting it now. I would probably have picked the default `name` to be simply `ignite`, because technically that is the "root" logger for the ignite library. The `.` separators in the logger name define a hierarchy. So if I want to modify all loggers used by ignite only, I can use the `logging.getLogger("ignite")` object in my application code. Was there another motivation for choosing `ignite.root.logger` instead?

I think it's good to have the `reset` flag default to `False`, even though it changes the behavior of `setup_logger()` compared to previous ignite versions. But as it's a pre-stable release version, this should certainly be fine with users. Otherwise, I think the `reset` flag may not actually have any meaningful use? I was mostly suggesting it in the first place to maintain existing behavior, but be able to alter it via opt-in. Having the `reset=False` behavior should be the new norm, though, so not sure there is any meaningful use case for `reset=True`. Application code should normally set up logging at the very start of the main function, after which there seems no need to reset. But fine if the option exists, of course.

Instead of a `reset` flag, another option may have been to split the functionality of `setup_logger()` into separate more reusable functions with each a clearer and more limited scope. To set the logging level, one should anyway just use the API provided by Python's logging module. I guess this mainly leaves a utility function to (re-)set the handlers depending on distributed_rank. If one were to remove `level` setting from `setup_logger()`, the function would indeed only be used to set the handlers as desired (as in `setup_logger_handlers()` kind of)?

Anyway, these were just some thoughts I had and missed to share earlier. It is looking fine as is for my use case.

**Comment by sdesrozis** [comment]
@schuhschuh Thank you for your feedback.

I realized that `ignite.root.Logger` was not clever and planed to rename to `ignite` as you suggested. I have been a bit distracted lately, so I haven't done it yet, my bad.

I do appreciate your idea to split the existing `setup_logger` to better use the `logging` configuration. We could implement the actual `setup_logger` considering this idea and see if `setup_logger` is still needed.

Finally, I agree that the `reset` flag could be useless and depends on the usage. Let's see if it should be deprecated in the future versions. Actually, it helps to get the previous behaviour as you mentioned.

@schuhschuh if you want to help on this topic, don't hesitate to create a new PR !! Thank you again!
